### PR TITLE
open_datatree performance improvement on NetCDF, H5, and Zarr files

### LINKED ISSUE #8994: Improving performance of open_datatree
### What is your issue?

The implementation of `open_datatree` works, but is inefficient, because it calls `open_dataset` once for every group in the file. We should refactor this to improve the performance, which would fix issues like https://github.com/xarray-contrib/datatree/issues/330.

We discussed this in the [datatree meeting](https://github.com/pydata/xarray/issues/8747), and my understanding is that concretely we need to:

- [ ] Create an asv benchmark for `open_datatree`, probably involving first writing then benchmarking the opening of a special netCDF file that has no data but lots of groups.
- [ ] Refactor the [`NetCDFDatastore`](https://github.com/pydata/xarray/blob/748bb3a328a65416022ec44ced8d461f143081b5/xarray/backends/netCDF4_.py#L319) class to only create one `CachingFileManager` object per file, not one per group, see https://github.com/pydata/xarray/blob/748bb3a328a65416022ec44ced8d461f143081b5/xarray/backends/netCDF4_.py#L406.
- [ ] Refactor `NetCDF4BackendEntrypoint.open_datatree` to use an implementation that goes through `NetCDFDatastore` without calling the top-level `xr.open_dataset` again.
- [ ] Check the performance of calling `xr.open_datatree` on a netCDF file has actually improved.

It would be great to get this done soon as part of the datatree integration project. @kmuehlbauer I know you were interested - are you willing / do you have time to take this task on?

### PR DESCRIPTION
**Pull request by aladinor**
open_datatree performance improvement on NetCDF files

- [x] Closes #8994 (NetCDF + Zarr datatree)
- [x] Tests added
- [x] User visible changes (including notable bug fixes) are documented in `whats-new.rst`

### CODE REVIEW

File: xarray/backends/zarr.py
Comment (shoyer) [inline xarray/backends/zarr.py:419]: Could you rewrite `open_group` to call `open_store()` internally? That would reduce the amount of duplicated code and make this easier to maintain going forward.

File: xarray/backends/netCDF4_.py
Comment (flamingbear) [reply xarray/backends/netCDF4_.py:19]: I think if you move these changes back to `backends/common` and leave them in `_open_datatree_netcdf`, You might be able to import the store for both the legacyhdf5 and the netcdf4 libraries. and then call `_open_datatree_netcdf` for both legacyhdf and netcdf.

currently `_open_datatree_netcdf`takes `ncDataset: ncDataset | ncDatasetLegacyH5,`

you might be able to include a new param with type `cdfDataStore: NetCDF4DataStore | H5NetCDFStore` and pass the appropritate one from both the n5netcdf_.py and netCDF4_.py backends.

**Review by flamingbear — COMMENTED** [review]
I had thoughts about the legacyhdf5 api and how it might be incorporated.

**Review by aladinor — COMMENTED** [review]
renaming variables

File: xarray/backends/netCDF4_.py
Comment (aladinor) [reply xarray/backends/netCDF4_.py:19]: Hi @flamingbear, 

Following up on your comments, I would like to ask: Does `datatree` support the `to_h5` method? I was looking for this method implementation or `test` but didn't find it. Therefore, do we need to support the `open_datatree` method for `h5` files even if we do not have a `to_h5`?

Please let me know your thoughts.

Comment (kmuehlbauer) [reply xarray/backends/netCDF4_.py:19]: @aladinor I think @flamingbear will have a much more detailed answer, so here just some short thoughts in advance. Xarray is able to use `netCDF4` and `h5netcdf` to read (almost) any HDF5 and netCDF4 files (a netCDF4 file is essentially an HDF5 file). `netcdf4-python` is able to read non-conforming files ( as well as h5netcdf). Same when writing, `engine="netcdf4"` and `engine="h5netcdf"` will create similar (netCDF4) files. But here files will be standard conforming (beside using engine `h5netcdf` with `invalid_netcdf=True`).

Comment (flamingbear) [reply xarray/backends/netCDF4_.py:19]: Yes, Kai is correct.  The h5netcdf is just a legacy/alternative interface to the netcdf4 files.  Both libraries supported a `Dataset` class with similar interfaces. The original datatree implementation [just chose which `Dataset` class](https://github.com/xarray-contrib/datatree/blob/0afaa6cc1d6800987d8b9c37a604dc0a8c68aeaa/datatree/io.py#L22-L34) to use based on which library was available and used them the same way.  Let me know if this helps or you want more info?

Comment (aladinor) [reply xarray/backends/netCDF4_.py:19]: Thanks, @kmuehlbauer and @flamingbear, for your feedback. I have one more question: Can I use the proposed implementation to open datatree stored in NetCDF4 files for H5 files?

Comment (flamingbear) [reply xarray/backends/netCDF4_.py:19]: Hi, I was just looking at this now.  I was hoping that we could pass just the class `NetCDF4DataStore` or `H5NetCDFStore` to the common function and just `GenericStore.open()`, but it looks like there are different keywords to the open function.  
I was just checking to see if in NetCDF4_.open_datatree you can open the store, and pass that and the class to a common open_datatree.

``` python
        filename_or_obj = _normalize_path(filename_or_obj)
        store = NetCDF4DataStore.open(...)
        return _open_datatree_netcdf_common(NetCDF4DataStore, store, ...rest)
```

And in h5netcdf_.open_datatree:

``` python

        filename_or_obj = _normalize_path(filename_or_obj)
        store = H5NetCDFStore.open(...)
        return _open_datatree_netcdf_common(H5NetCDFStore, store, ...rest)
```

though, I haven't got it completed yet or determined if the different open() functions could be aligned with the same keywords.

Comment (aladinor) [reply xarray/backends/netCDF4_.py:19]: Thanks, @flamingbear, for taking a look at it. I also feel more confident about keeping them separately, as we must create a `group_store` when looping through each node.

```python 
for group in _iter_nc_groups(store.ds):
            group_path = str(parent / group[1:])
            group_store = NetCDF4DataStore(manager, group=group_path, **kwargs) #  <--- here
            store_entrypoint = StoreBackendEntrypoint()
```

 I created separate implementations to ensure all keywords are aligned with each `store` object.

File: xarray/backends/zarr.py
Comment (aladinor) [reply xarray/backends/zarr.py:1205]: Thanks, @keewis, for bringing this up. We added the group parameter in case someone decides just to open a specific group within the datatree (e.g. `open_datatree(path2tree, group='group_0/subgroup_1'`). I will define this as `str | Iterable[str] | callable`

Comment (keewis) [reply xarray/backends/zarr.py:1205]: if it turns out to be too difficult, we can also aim for `group: str | None = None`, and add `Iterable[str] | Callable` in a later PR (we would need additional code and tests for that new feature)

Comment (aladinor) [reply xarray/backends/zarr.py:1205]: I got a Mypy error. I am just rolling back to the original idea and then we can add a new PR in the future.

Comment (aladinor) [reply xarray/backends/zarr.py:1205]: @keewis I will retry with the open_datatree function for h5 files to see if it works. I feel like I tested it using `None`, but I'm not sure. Let me see if I was making a mistake.

Comment (aladinor) [reply xarray/backends/zarr.py:1205]: I ran the following command locally to see what was happening 

```python
 python -m mypy --follow-imports=skip xarray/backends/h5netcdf_.py 
```
and this is the output 

```python 
xarray/backends/h5netcdf_.py:477: error: Value of type "str | Iterable[str] | Callable[..., Any] | None" is not indexable  [index] 
Found 2 errors in 1 file (checked 1 source file)
```
seems like something is happening in this line

```python
group_path = str(parent / group[1:])
```
I think it is because I am using  `group[1:]`  slicing. I need to concatenate the parent node with the subgroup. This subgroup is a string that starts with `/` (e.g. '/subgroup_1'), so if I use the whole string, it won't allow string concatenation. Any thoughts how to handle this?

Comment (flamingbear) [reply xarray/backends/zarr.py:1205]: > open_datatree function for h5 files to see if it works

Remember, these are actually still netcdf4 files, just using a different library to access. <https://github.com/aladinor/xarray/blob/datatree-zarr/xarray/backends/h5netcdf_.py#L157-L164>

I'm still looking to see if the open_datatree can be simplified, trying to understand where all of the keywords came from.

File: xarray/backends/h5netcdf_.py
Comment (flamingbear) [inline xarray/backends/h5netcdf_.py:433]:
```suggestion
	*,
        mask_and_scale=True,
```
separate positional from keyword args like you do in the netcdf file.

File: xarray/backends/zarr.py
Comment (aladinor) [reply xarray/backends/zarr.py:1205]: @keewis and @flamingbear, I think I solved the issue with type hints. It was caused by my using `group[1:]` slicing notation. I refactored it and submitted a new commit.

File: xarray/backends/netCDF4_.py
Comment (flamingbear) [reply xarray/backends/netCDF4_.py:19]: I'm trying not to harp on this too much, but that `group_store` you create has the same arguments whether it's a `NetCDF4DataStore` or a `H5NetCDFStore` which could be passed as a generic store.  I'm still on the fence if this is easier to maintain because it's in one place or if it's just adding complexity to try to reuse the code.

File: xarray/backends/h5netcdf_.py
Comment (flamingbear) [inline xarray/backends/h5netcdf_.py:441]:
```suggestion
	mode="r",
        format=None,
```
I am still trying to figure out how you gathered these keyword args, but the h5netcdf store's open, takes a mode.

File: xarray/backends/h5netcdf_.py
Comment (flamingbear) [inline xarray/backends/h5netcdf_.py:460]:
```suggestion
            mode=mode,
            format=format,
```

File: xarray/backends/netCDF4_.py
Comment (flamingbear) [reply xarray/backends/netCDF4_.py:19]: I'll let someone else make this call. 

The only diffs I found in the Store.open() keywords are:

 - netcdf:
   `clobber`, `diskless`, `persist`, `autoclose`

- h5netcdf:
   `invalid_netcdf`, `phony_dims`, `decode_vlen_strings`, `driver`, `driver_kwds`
   
   But I'm a little naive about the named keywords vs **kwargs, and how I'd reconcile those.

File: xarray/backends/h5netcdf_.py
Comment (flamingbear) [reply xarray/backends/h5netcdf_.py:441]: @aladinor This was this only thing remaining for me.  And we're talking about it now.  I'm fine with the separate backends. Looks good.

**Review by flamingbear — APPROVED** [review]
This looks great. Thanks for working through the dual library stuff with me.

**Comment by TomNicholas** [comment]
Yes very excited by this! Two final things:
- This deserves a `whats-new.rst` entry!
- Would you be willing to add an benchmark test? You can see here how we benchmark opening and loading a single netCDF file

https://github.com/pydata/xarray/blob/447e5a3d16764a880387d33d0b5938393e167817/asv_bench/benchmarks/dataset_io.py#L125

Alternatively we could leave adding that benchmark to a separate PR?

**Comment by kmuehlbauer** [comment]
Sorry for being late to the party, but do we really want to have `mode` in `open_datatree`? The `open_*` functions are for read access only (imho).

**Review by kmuehlbauer — APPROVED** [review]
Beside my other comment on the `open_*`-function kwargs there is not much to add to @TomNicholas's [comment](https://github.com/pydata/xarray/pull/9014#issuecomment-2148526504). 

We might want to further deduplicate code in the backends (by moving this into backend/store.py). There has already quite some work been done by @jthielen in #7437.

Should `open_datatree` be added to the API with this PR too?

Those issues can be handled in subsequent PR.

Great work @aladinor :rocket:.

**Comment by keewis** [comment]
In the meeting yesterday we decided to not bother with deduplication for now (and anyways it is nice to have backends somewhat self-contained).

The API will be extended in #9033, the idea is to have one single PR that marks the entire `DataTree` API as public.

**Review by TomNicholas — CHANGES_REQUESTED** [review]
This is very close! We should try to get it in.

> This deserves a whats-new.rst entry!

We don't need to wait for anything here - one of the maintainers can just add this by pushing to this branch.

> Would you be willing to add an benchmark test? You can see here how we benchmark opening and loading a single netCDF file

I'm fine to leave this to a later PR, we should just create an issue so we remember to add it later. The main purpose is as a performance regression test - we already know this PR is a big improvement!

We shouldn't be adding extra kwargs though - as far as I can tell `open_datatree` should have exactly the same signature as `open_dataset`. See [Kai's comment](https://github.com/pydata/xarray/pull/9014#issuecomment-2148958517). Once that's done I'm happy to approve and merge.

**Comment by aladinor** [comment]
Hi @TomNicholas, 

Thanks for your review and my apologies for not joining today's meeting.  My internet connection is quite limited, which is causing some difficulties in staying fully connected. 

I will ensure to have some of this addressed by our upcoming meeting on 06/18. However,  I have some comments to share.
> This is very close! We should try to get it in.
> 
> > This deserves a whats-new.rst entry!
> 
> We don't need to wait for anything here - one of the maintainers can just add this by pushing to this branch.
> 
I can work on this for our next meeting. It won't take to much work. 
> > Would you be willing to add an benchmark test? You can see here how we benchmark opening and loading a single netCDF file
> 
I may be able to work on it, but I'm not sure if it will be ready for next week. Let me check it out
> I'm fine to leave this to a later PR, we should just create an issue so we remember to add it later. The main purpose is as a performance regression test - we already know this PR is a big improvement!
> 
> We shouldn't be adding extra kwargs though - as far as I can tell `open_datatree` should have exactly the same signature as `open_dataset`. See [Kai's comment](https://github.com/pydata/xarray/pull/9014#issuecomment-2148958517). Again @owenlittlejohns or @flamingbear you can just push to this branch to push it over the finish line. Once that's done I'm happy to approve and merge.

Regarding this last point, I have been using the `open_dataset` parameters since the beginning. However, I completely agree with you and @kmuehlbauer about removing those unnecessary parameters. This will be ready as soon as possible.

**Comment by flamingbear** [comment]
This looks to be updated, and I would like to try to merge it before next week. The items remaining from @TomNicholas [comment](https://github.com/pydata/xarray/pull/9014#pullrequestreview-2111487817) are addressed but for the benchmark test that can be done in a separate PR.

The keywords are not exact with the open_dataset, but have removed those related to writing.  That seems fine to me, but looking for second opinions via thumbs up. Otherwise I will add a `plan to merge` label and wait.

**Comment by dcherian** [comment]
Nice! Merging since this has been thoroughly looked at.

**Comment by mgrover1** [comment]
Wooooo great work all!!

**Comment by aladinor** [comment]
I was nice to work in this PR. See you all in the next datatree meeting.

**Comment by shoyer** [comment]
Great work Alfonso!

**Pull request merged**
